### PR TITLE
Fixed duplicate properties creation

### PR DESCRIPTION
**Pull request by dilpreet92**
Fix for duplicate creation of properties

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://dilpreet92-fix-property.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://dilpreet92-fix-property.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/app/models/spree/product_property.rb [hunk 1]
```diff
@@ -7,7 +7,8 @@ class ProductProperty < Spree::Base
       belongs_to :property, class_name: 'Spree::Property'
     end
 
-    validates :property, presence: true
+    validates :product, presence: true
+    validates :property, uniqueness: { scope: :product }, presence: true
```
Comment (priyank-gupta) [inline core/app/models/spree/product_property.rb]: @dilpreet92 We should add `message: :already_linked` with uniqueness validation.

File: core/app/models/spree/property.rb [hunk 2]
```diff
@@ -6,7 +6,8 @@ class Property < Spree::Base
     has_many :product_properties, dependent: :delete_all, inverse_of: :property
     has_many :products, through: :product_properties
 
-    validates :name, :presentation, presence: true
+    validates :presentation, presence: true
+    validates :name, uniqueness: { scope: :presentation }, presence: true
```
Comment (priyank-gupta) [inline core/app/models/spree/property.rb]: @dilpreet92 I think we can add uniqueness on name but not sure why we should add presentation scope.

Comment (dilpreet92) [inline core/app/models/spree/property.rb]: @priyank-gupta There might be a case when property name of the product is registered but it's presentation might be different

**Comment by Mafi88** [comment]
@dilpreet92 @priyank-gupta  I also agree that name validation shouldn't take ```presentation``` into consideration. Name should be unique, so it could be used like ```slug``` in ```Product``` (no hardcoded id's when there is need to find specific property).

Could you also add validation specs for this change?

And I'm not sure if it won't be problematic for existig spree applications that don't have unique properties? Maybe there should be some migration? What do you think @damianlegawiec, @jitendra?

**Comment by priyank-gupta** [comment]
@Mafi88 Agreed. Keeping backward compatibility in mind some kind of migration should be there.
Like number countering duplicate names.

**Comment by dilpreet92** [comment]
@Mafi88 - Done with the fix. What do you suggest for keeping backward compatibility?

**Comment by Mafi88** [comment]
@dilpreet92 No matter what we do, if name was used in some operation (like searching), this change will break this operation.
So this will require more attention than usual update. Therefore I think it should be added in some major release. Maybe 3.1? What do you think @damianlegawiec @priyank-gupta @jitendra?

I consider two scenarios:
- leave it as it is - there will be validation errors after migration to new version and it will force developers to change names according to their needs
- create a migration that will change names to unique (so there won't be any problems in applications that are not using name)

**Comment by damianlegawiec** [comment]
Closing all uniqueness validation PRs in favour of one big PR doing it right (DB level + app level) for 3.2 release.

**Pull request closed without merging**